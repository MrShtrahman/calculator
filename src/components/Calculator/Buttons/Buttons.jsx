import "./Buttons.css";

import CalcButton from "./CalcButton/CalcButton";
import React from "react";
import { useOnClick } from "../../../hooks/useOnClick";

const chooseColor = value => {
  if (!isNaN(value)) return "blue";
  else {
    switch (value) {
      case "C":
        return "red";
      default:
        return "purple";
    }
  }
};

const Buttons = () => {
  const onClick = useOnClick();

  const values = [
    ["M+", "M-", "MR", "MC"],
    [7, 8, 9, "+"],
    [4, 5, 6, "-"],
    [1, 2, 3, "*"],
    ["C", 0, "=", "/"]
  ];

  return values.map((row, index) => (
    <div className="buttonsRow" key={index}>
      {row.map(value => (
        <CalcButton
          key={value}
          value={value}
          color={chooseColor(value)}
          clicked={() => onClick(value)}
        />
      ))}
    </div>
  ));
};

export default Buttons;
