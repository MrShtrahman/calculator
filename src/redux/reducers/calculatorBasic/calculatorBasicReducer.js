import { calculatorBasic } from '../../actions/actionTypes';

const initialState = {
    display: '0',
    left: 0,
    right: 0,
    operator: null
}

const calculatorReducer = (state = initialState, action) => {
    switch(action.type) {
        case calculatorBasic.SET_DISPLAY :
            return {
                ...state,
                display: action.payload
            }
        case calculatorBasic.SET_LEFT :
            return {
                ...state,
                left: action.payload,
                display: action.payload + ''
            }
        case calculatorBasic.SET_RIGHT :
            return {
                ...state,
                right: action.payload,
                display: action.payload + ''
            }
        case calculatorBasic.SET_OPERATOR: 
            return {
                ...state,
                operator: action.payload
            }
        case calculatorBasic.CLEAR: 
            return {
                ...state,
                display: '0',
                operator: null,
                left: 0,
                right: 0
            }
        default:
            return state
    }
}

export default calculatorReducer;