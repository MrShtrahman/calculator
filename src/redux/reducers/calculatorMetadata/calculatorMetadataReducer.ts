import * as actions from "../../actions/calculatorMetadata";

import { ActionType, getType } from "typesafe-actions";

export type MetadataAction = ActionType<typeof actions>;
const initialState = {
  isLeft: true,
  equalClicked: false,
  operatorClicked: false,
  history: ""
};

export type MetadataState = {
  isLeft: boolean;
  equalClicked: boolean;
  operatorClicked: boolean;
  history: string;
};

export const calculatorMetadataReducer = (
  state: MetadataState = initialState,
  action: MetadataAction
): MetadataState => {
  switch (action.type) {
    case getType(actions.setEqualClicked):
      return { ...state, equalClicked: action.payload ? true : false };
    case getType(actions.setOperatorClicked):
      return { ...state, operatorClicked: action.payload ? true : false };
    case getType(actions.setIsLeft):
      return { ...state, isLeft: action.payload ? true : false };
    case getType(actions.addToHistory):
      return { ...state, history: action.payload.toString() };
    case getType(actions.resetHistory):
      return { ...state, history: "" };
    default:
      return state;
  }
};
