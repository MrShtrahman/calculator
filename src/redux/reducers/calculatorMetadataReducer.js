import { calculatorMetadata } from '../actions/actionTypes'

const initialState = {
    isLeft: true,
    equalClicked: false,
    operatorClicked: false
}

const calculatorMetadataReducer = (state = initialState, action) => {
    switch(action.type) {
        case calculatorMetadata.SET_EQUAL_CLICKED:
            return {
                ...state,
                equalClicked: action.payload
            }
        case calculatorMetadata.SET_OPERATOR_CLICKED:
            return {
                ...state,
                operatorClicked: action.payload
            }
        case calculatorMetadata.SET_IS_LEFT:
            return {
                ...state,
                isLeft: action.payload
            }
        default: return state
    }
}

export default calculatorMetadataReducer