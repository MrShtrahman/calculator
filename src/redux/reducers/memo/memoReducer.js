import { memo } from '../../actions/actionTypes';

const initialState = {
    memo: 0,
    isMemoShown: false
}

const memoReducer = (state = initialState, action) => {
    switch(action.type) {
        case memo.ADD_TO_MEMO: 
            return {
                memo: state.memo + action.payload,
                isMemoShown: true
            }
        case memo.SUB_FROM_MEMO: 
            return {
                memo: state.memo - action.payload,
                isMemoShown: true
            }
        case memo.RESET_MEMO: 
            return initialState
        default: return state
    }
}

export default memoReducer