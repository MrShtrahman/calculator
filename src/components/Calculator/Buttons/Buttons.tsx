import React, { ReactText } from "react";
import { isNumber } from "util";
import "./Buttons.css";
import CalcButton from "./CalcButton/CalcButton";
import { useOnClick } from "../../../hooks/useOnClick";

const chooseColor = (value: ReactText) => {
  if (isNumber(value)) return "blue";
  else {
    switch (value) {
      case "C":
        return "red";
      default:
        return "purple";
    }
  }
};

const Buttons = () => {
  const onClick = useOnClick();
  const values = [
    ["M+", "M-", "MR", "MC"],
    [7, 8, 9, "+"],
    [4, 5, 6, "-"],
    [1, 2, 3, "*"],
    ["C", 0, "=", "/"]
  ];

  return values.map((row, index) => (
    <div className="buttonsRow" key={index}>
      {row.map(value => (
        <CalcButton
          key={value}
          value={value}
          color={chooseColor(value)}
          onClick={() => onClick(value)}
          style={{
            height: "65px",
            width: "65px"
          }}
          size={"large"}
        />
      ))}
    </div>
  ));
};

export default Buttons;
