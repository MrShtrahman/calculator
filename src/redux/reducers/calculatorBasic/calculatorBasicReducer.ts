import * as actions from "../../actions/calculatorBasic";

import { ActionType, getType } from "typesafe-actions";

export type CalcBasicAction = ActionType<typeof actions>;

export type CalcBasicState = {
  display: string;
  left: number;
  right: number;
  operator: string;
};

const initialState = {
  display: "0",
  left: 0,
  right: 0,
  operator: "",
};

export const calculatorBasicReducer = (
  state: CalcBasicState = initialState,
  action: CalcBasicAction
): CalcBasicState => {
  switch (action.type) {
    case getType(actions.setRight):
      return {
        ...state,
        right: Number(action.payload),
        display: action.payload.toString(),
      };
    case getType(actions.setLeft):
      return {
        ...state,
        left: Number(action.payload),
        display: action.payload.toString(),
      };
    case getType(actions.setDisplay):
      return { ...state, display: action.payload.toString() };
    case getType(actions.setOperator):
      return { ...state, operator: action.payload.toString() };

    case getType(actions.clear):
      return initialState;

    default:
      return state;
  }
};
