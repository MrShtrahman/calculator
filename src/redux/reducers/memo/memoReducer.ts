import * as actions from "../../actions/memo";

import { ActionType, getType } from "typesafe-actions";

export type MemoAction = ActionType<typeof actions>;

export type MemoState = {
  memo: number;
  isMemoShown: boolean;
};

const initialState = {
  memo: 0,
  isMemoShown: false
};

export const memoReducer = (
  state: MemoState = initialState,
  action: MemoAction
): MemoState => {
  switch (action.type) {
    case getType(actions.addToMemo):
      return {
        memo: state.memo + Number(action.payload),
        isMemoShown: true
      };
    case getType(actions.subFromMemo):
      return {
        memo: state.memo - Number(action.payload),
        isMemoShown: true
      };
    case getType(actions.resetMemo()):
      return initialState;
    default:
      return state;
  }
};
