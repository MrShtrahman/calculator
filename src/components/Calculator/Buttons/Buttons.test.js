import { combineReducers, createStore } from 'redux';
import { configure, mount } from 'enzyme';

import Adapter from 'enzyme-adapter-react-16';
import Buttons from './Buttons';
import { Provider } from 'react-redux';
import React from 'react';
import calculatorBasicReducer 
    from '../../../redux/reducers/calculatorBasic/calculatorBasicReducer';
import calculatorMetadataReducer 
    from '../../../redux/reducers/calculatorMetadata/calculatorMetadataReducer';
import expect from 'expect'
import memoReducer from '../../../redux/reducers/memo/memoReducer';

configure({
    adapter: new Adapter()
})

describe('<Buttons /> component', () => {
    const reducer = combineReducers({
        calculatorBasicReducer,
        calculatorMetadataReducer,
        memoReducer
    })
    const mockStore = createStore(reducer, {
        memoReducer: {},
        calculatorBasicReducer: {},
        calculatorMetadataReducer: {}
    })

    it('renders', () => {
        const wrapper = mount(
            <Provider store = {mockStore}>
                <Buttons />
            </Provider>
        );
        
        expect(wrapper).toMatchSnapshot()
    })
})