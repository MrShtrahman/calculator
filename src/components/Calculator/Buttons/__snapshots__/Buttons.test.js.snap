// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Buttons /> component renders 1`] = `ReactWrapper {}`;
